File: backend/src/main/java/com/alantek/caja/modulo/email/service/EmailConfigService.java
package com.alantek.caja.modulo.email.service;

import com.alantek.caja.modulo.email.dto.EmailConfiguracionRequest;
import com.alantek.caja.modulo.email.dto.EmailConfiguracionResponse;
import com.alantek.caja.modulo.email.dto.EmailPlantillaRequest;
import com.alantek.caja.modulo.email.dto.EmailPlantillaResponse;
import com.alantek.caja.modulo.email.entity.EmailConfiguracion;
import com.alantek.caja.modulo.email.entity.EmailPlantilla;
import com.alantek.caja.modulo.email.repository.EmailConfiguracionRepository;
import com.alantek.caja.modulo.email.repository.EmailPlantillaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class EmailConfigService {

    private final EmailConfiguracionRepository configRepo;
    private final EmailPlantillaRepository plantillaRepo;

    public EmailConfigService(EmailConfiguracionRepository configRepo, EmailPlantillaRepository plantillaRepo) {
        this.configRepo = configRepo;
        this.plantillaRepo = plantillaRepo;
    }

    // ── Configuracion ─────────────────────────────────────────────────────

    public Optional<EmailConfiguracionResponse> obtenerConfiguracion() {
        return configRepo.findFirstByActivoTrue().or(() -> configRepo.findAll().stream().findFirst())
                .map(this::toConfigResponse);
    }

    @Transactional
    public EmailConfiguracionResponse guardarConfiguracion(EmailConfiguracionRequest request) {
        EmailConfiguracion config = configRepo.findFirstByActivoTrue().orElseGet(EmailConfiguracion::new);
        config.setMetodo(request.metodo());
        config.setSmtpHost(request.smtpHost());
        config.setSmtpPort(request.smtpPort());
        config.setSmtpUsername(request.smtpUsername());
        config.setSmtpPassword(request.smtpPassword());
        config.setSmtpUseTls(request.smtpUseTls() != null ? request.smtpUseTls() : true);
        config.setSmtpUseSsl(request.smtpUseSsl() != null ? request.smtpUseSsl() : false);
        config.setApiUrl(request.apiUrl());
        config.setApiKey(request.apiKey());
        config.setApiProvider(request.apiProvider());
        config.setFromEmail(request.fromEmail());
        config.setFromName(request.fromName());
        config.setActivo(request.activo() != null ? request.activo() : true);
        return toConfigResponse(configRepo.save(config));
    }

    @Transactional
    public Map<String, Object> eliminarConfiguracion(Long id) {
        configRepo.deleteById(id);
        return Map.of("message", "Configuracion eliminada");
    }

    // ── Plantillas ────────────────────────────────────────────────────────

    public List<EmailPlantillaResponse> listarPlantillas(String modulo) {
        if (modulo != null && !modulo.isBlank()) {
            return plantillaRepo.findByModuloOrderByNombreAsc(modulo).stream().map(this::toPlantillaResponse).toList();
        }
        return plantillaRepo.findAllByOrderByModuloAscNombreAsc().stream().map(this::toPlantillaResponse).toList();
    }

    public Optional<EmailPlantillaResponse> obtenerPlantilla(Long id) {
        return plantillaRepo.findById(id).map(this::toPlantillaResponse);
    }

    @Transactional
    public EmailPlantillaResponse guardarPlantilla(EmailPlantillaRequest request) {
        EmailPlantilla plantilla = new EmailPlantilla();
        plantilla.setModulo(request.modulo());
        plantilla.setNombre(request.nombre());
        plantilla.setAsunto(request.asunto());
        plantilla.setCuerpoHtml(request.cuerpoHtml());
        plantilla.setVariables(request.variables());
        plantilla.setActivo(request.activo() != null ? request.activo() : true);
        return toPlantillaResponse(plantillaRepo.save(plantilla));
    }

    @Transactional
    public EmailPlantillaResponse actualizarPlantilla(Long id, EmailPlantillaRequest request) {
        EmailPlantilla plantilla = plantillaRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Plantilla no encontrada"));
        plantilla.setModulo(request.modulo());
        plantilla.setNombre(request.nombre());
        plantilla.setAsunto(request.asunto());
        plantilla.setCuerpoHtml(request.cuerpoHtml());
        plantilla.setVariables(request.variables());
        plantilla.setActivo(request.activo() != null ? request.activo() : true);
        return toPlantillaResponse(plantillaRepo.save(plantilla));
    }

    @Transactional
    public Map<String, Object> eliminarPlantilla(Long id) {
        plantillaRepo.deleteById(id);
        return Map.of("message", "Plantilla eliminada");
    }

    @Transactional
    public EmailPlantillaResponse togglePlantilla(Long id) {
        EmailPlantilla plantilla = plantillaRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("Plantilla no encontrada"));
        plantilla.setActivo(!plantilla.getActivo());
        return toPlantillaResponse(plantillaRepo.save(plantilla));
    }

    // ── Mappers ───────────────────────────────────────────────────────────

    private EmailConfiguracionResponse toConfigResponse(EmailConfiguracion c) {
        return new EmailConfiguracionResponse(
                c.getId(), c.getMetodo(), c.getSmtpHost(), c.getSmtpPort(),
                c.getSmtpUsername(), c.getSmtpUseTls(), c.getSmtpUseSsl(),
                c.getApiUrl(), c.getApiProvider(), c.getFromEmail(), c.getFromName(), c.getActivo()
        );
    }

    private EmailPlantillaResponse toPlantillaResponse(EmailPlantilla p) {
        return new EmailPlantillaResponse(
                p.getId(), p.getModulo(), p.getNombre(), p.getAsunto(),
                p.getCuerpoHtml(), p.getVariables(), p.getActivo()
        );
    }
}

File: frontend/src/app/core/services/email-config.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { EmailConfiguracion, EmailPlantilla, EmailConfiguracionRequest, EmailPlantillaRequest } from '../models/email.model';

@Injectable({ providedIn: 'root' })
export class EmailConfigService {
  private readonly base = `${environment.apiUrl}/config/email`;

  constructor(private readonly http: HttpClient) {}

  obtenerConfiguracion(): Observable<EmailConfiguracion | null> {
    return this.http.get<EmailConfiguracion | null>(`${this.base}/configuracion`);
  }

  guardarConfiguracion(request: EmailConfiguracionRequest): Observable<EmailConfiguracion> {
    return this.http.post<EmailConfiguracion>(`${this.base}/configuracion`, request);
  }

  listarPlantillas(modulo?: string): Observable<EmailPlantilla[]> {
    let params = new HttpParams();
    if (modulo) params = params.set('modulo', modulo);
    return this.http.get<EmailPlantilla[]>(`${this.base}/plantillas`, { params });
  }

  obtenerPlantilla(id: number): Observable<EmailPlantilla> {
    return this.http.get<EmailPlantilla>(`${this.base}/plantillas/${id}`);
  }

  crearPlantilla(request: EmailPlantillaRequest): Observable<EmailPlantilla> {
    return this.http.post<EmailPlantilla>(`${this.base}/plantillas`, request);
  }

  actualizarPlantilla(id: number, request: EmailPlantillaRequest): Observable<EmailPlantilla> {
    return this.http.put<EmailPlantilla>(`${this.base}/plantillas/${id}`, request);
  }

  eliminarPlantilla(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.base}/plantillas/${id}`);
  }

  togglePlantilla(id: number): Observable<EmailPlantilla> {
    return this.http.put<EmailPlantilla>(`${this.base}/plantillas/${id}/toggle`, null);
  }

  testEmail(to: string): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(`${this.base}/test`, { to });
  }

  listarModulos(): Observable<string[]> {
    return this.http.get<string[]>(`${this.base}/modulos`);
  }
}
